proposals/20210525-exmplefilter.md

```diff
@@ -0,0 +1,70 @@
+#### SIG TFX-Addons
+# Project Proposal
+## Instructions
+Copy this template, naming it with the date in `yyyymmdd` format followed by a short name, and fill in the blanks below.
+When you're ready for review, place it in `/proposals` and create a pull request.
+
+---
+
+**Your name:** Ryan Clough
+
+**Your email:** rclough@spotify.com
+
+**Your company/organization:** Spotify
+
+**Project name:** Example Filter
+
+## Project Description
+Beam bassed component that can filter Examples based on a user-defined predicate function.
+
+## Project Category
+Choose 1: Component
+
+## Project Use-Case(s)
+Data can be imported into TFX in a number of ways, and indeed, sometimes the dataset you wish to load is not under your direct 
+control. In cases like these, it is useful to have a component that can filter your input data with simple rules. Ex: filter 
+all records where `feature_a >= 1`. 
+
+Our organization currently has a component for this purpose that is in active use. It is not ass robust as it could be.
+
+It is also worth conidering that we may wish to try and promote this functionality to be included in the TFX core base ExampleGen,
+so that the filtering could be done within any ExampleGen based component.
+
+## Project Implementation
+Spotify can provide the current implementation, which is based off of an old version of Tensorflow Transform. At a high level, use
+of the coomponent looks like:
+
+```python
+def perdicate_fn(example)
+    # Throw out Examples that used a credit card
+    if b'Credit Card' in example['payment_type']:
+        return False
+    return True
+...
+
+filtered_examples = ExampleFilter(
+    examples=examples.output,
+    schema=schema.output,
+    module_file=filter_module,
+)
+```
+
+The current implementation is a bit dated and not so robust. It depends on a deprecated TFT proto coder, and only works on
+TF Records, as it does not make use of TFXIO. As part of bringing this to TFX-addons, I think it is worth iterating on the
+current design. Some initial ideas for change might be:
+
+* Implementing it more flexibly in TFXIO
+* Determine if there's a way to implement it without requiring a schema
```

Comment:
Why don't we use pa.RecordBatch as example instead?  And simplify it. This may speed it up and simplify the component itself given that TFXIO generates Beam Source of RecordBatch

Reply:
That sounds on the surface like a decent idea; I'm not very good at beam so I can't make much argument for or against it.  While back I tried converting it to TFXIO but gave up after time boxing it, you can see the results of this effort in this public TFX thread: https://groups.google.com/a/tensorflow.org/g/tfx/c/nq9rWPAmTXw/m/eCiTISXrFAAJ

particularly this statement:

> At a high level, there's currently no support for turning a RecordBatch back to tf.Examples. We do have a piece of code that does that, but it's not wired in (i.e. not available as a python API). It may be made available in the future.